packages/ui/VirtualList/VirtualListBase.js

```diff
@@ -23,728 +27,771 @@ const gridListItemSizeShape = PropTypes.shape({
-			style: {...itemElement.props.style, ...style}
-		});
-	}
-
-	applyStyleToHideNode = (index) => {
-		const key = index % this.state.numOfItems;
-		this.cc[key] = <div key={key} style={{display: 'none'}} />;
-	}
-
-	positionItems ({updateFrom, updateTo}) {
-		const
-			{dataSize} = this.props,
-			{isPrimaryDirectionVertical, dimensionToExtent, primary, secondary, scrollPosition} = this;
-		let
-			{primaryPosition, secondaryPosition} = this.getGridPosition(updateFrom),
-			hideTo = 0,
-			width, height;
-
-		primaryPosition -= scrollPosition;
-		width = (isPrimaryDirectionVertical ? secondary.itemSize : primary.itemSize) + 'px';
-		height = (isPrimaryDirectionVertical ? primary.itemSize : secondary.itemSize) + 'px';
-
-		if (updateTo > dataSize) {
-			hideTo = updateTo;
-			updateTo = dataSize;
+
+		getItemNode = (index) => {
+			const ref = this.itemContainerRef;
+
+			return ref ? ref.children[index % this.state.numOfItems] : null;
 		}
 
-		// positioning items
-		for (let i = updateFrom, j = updateFrom % dimensionToExtent; i < updateTo; i++) {
-			if (this.updateFrom === null || this.updateTo === null || this.updateFrom > i || this.updateTo <= i) {
-				this.applyStyleToNewNode(i, width, height, primaryPosition, secondaryPosition);
-			} else {
-				this.applyStyleToExistingNode(i, width, height, primaryPosition, secondaryPosition);
-			}
+		getXY = (isPrimaryDirectionVertical, primaryPosition, secondaryPosition) => (isPrimaryDirectionVertical ? {x: secondaryPosition, y: primaryPosition} : {x: primaryPosition, y: secondaryPosition})
 
-			if (++j === dimensionToExtent) {
-				secondaryPosition = 0;
-				primaryPosition += primary.gridSize;
-				j = 0;
-			} else {
-				secondaryPosition += secondary.gridSize;
+		composeStyle (style, width, height, primaryPosition, secondaryPosition) {
+			const {x, y} = this.getXY(this.isPrimaryDirectionVertical, primaryPosition, secondaryPosition);
+
+			if (this.isItemSized) {
+				style.width = width;
+				style.height = height;
 			}
-		}
+			style.position = 'absolute';
 
-		for (let i = updateTo; i < hideTo; i++) {
-			this.applyStyleToHideNode(i);
+			/* FIXME: RTL / this calculation only works for Chrome */
+			style.transform = 'translate(' + (this.props.rtl ? -x : x) + 'px,' + y + 'px)';
 		}
 
-		this.updateFrom = updateFrom;
-		this.updateTo = updateTo;
-	}
+		applyStyleToNewNode = (index, ...rest) => {
+			const
+				{itemRenderer, getComponentProps, data} = this.props,
+				{numOfItems} = this.state,
+				key = index % numOfItems,
+				itemElement = itemRenderer({
+					data,
+					index,
+					key
+				}),
+				style = {},
+				componentProps = getComponentProps && getComponentProps(index) || {};
+
+			this.composeStyle(style, ...rest);
+
+			this.cc[key] = React.cloneElement(itemElement, {
+				...componentProps,
+				className: classNames(css.listItem, itemElement.props.className),
+				['data-preventscrollonfocus']: true, // Added this attribute to prevent scroll on focus by browser
```

Comment:
I'm not sure it would be better to ad `data-preventscrollonfocus` for `VirtualList` JS. But it would be ok now.

Reply:
`data-preventscrollonfocus` is a guard for unintended scrolling by focus() API call and not harmful to JS version. When browsers are ready to handle _focus option_ specified in https://html.spec.whatwg.org/multipage/interaction.html#dom-focus, `data-preventscrollonfocus` should be gone for both versions of lists.